torch/functional.py

```diff
@@ -71,6 +71,49 @@ def broadcast_tensors(*tensors):
     return _VF.broadcast_tensors(tensors)  # type: ignore
 
 
+def broadcast_shape(*shapes):
+    r"""
+
+    Similar to :func:`broadcast_tensors` but for shapes.
+
+    This is roughly equivalent to
+    `torch.broadcast_tensors(*map(torch.empty, shapes)).shape`
+    but avoids the need create to intermediate tensors. This is useful for
+    broadcasting tensors of common batch shape but different rightmost shape,
+    e.g. to broadcast mean vectors with covariance matrices.
+
+    Example::
+
+        >>> mean = torch.randn(30, 3)
+        >>> cov = torch.eye(20, 1, 3, 3)
+        >>> batch_shape = torch.broadcast_shape(mean.shape[:-1], cov.shape[:-2])
+        >>> batch_shape
+        torch.Size([20, 30])
+        >>> a = mean.expand(batch_shape + (-1,))
+        >>> b = cov.expand(batch_shape + (-1, -1))
+
+    Args:
+        \*shapes (torch.Size): Shapes of tensors.
+
+    Returns:
+        shape (torch.Size): A shape compatible with all input shapes.
+
+    Raises:
+        RuntimeError: If shapes are incompatible.
+    """
+    result = [1] * max(map(len, shapes))
+    for shape in shapes:
+        for i in range(-1, -1 - len(shape), -1):
+            if shape[i] == 1 or shape[i] == result[i]:
```

Comment:
@albanD @fritzo I am pretty sure tracing issue comes from Tensor-to-bool conversion here. I am not sure if it is possible to fix that in python, I am not very familiar with the codebase.

Reply:
I'm not sure moving to c++ would help.
Also this is just working with sizes right? Not Tensor?

Reply:
Here we're working with sizes, but torch.Size.__getitem__ returns torch.Tensor with zero dimension IIRC.
The point here is that we can pass tensor with batch shape 1 as tracer argument, which will select `continue` control flow, and during trace replay with tensor whose batch shape is greater than one, this expansion will not trigger, leading to wrong resulting shape.
Same applies to exception raising few lines below.

Reply:
It seems that adding this function to TorchScript might work... However I am not sure if it will work with shapes.

Reply:
Ho i see, because the function does not use Tensors at all, it does nothing from the point of view of tracing and the size that is computed is just used as a constant.
While the old code that was using only Tensor based functions did not had this issue.

I could see two options here:
- special case the MultivariateNormal code to use the more efficient code only if you're not tracing
- make this work with tracing, but I don't think that would be possible...

@suo might have a better idea?

Reply:
It is not exactly constant, seems like tracing accounts for sizes as well:
```
def foo(a):
    return a.shape[0]
traced = torch.jit.trace(foo, (torch.ones(3,4,5),))
traced(torch.ones(17))
>>>: tensor(17)
traced(torch.zeros(7,6,5))
>>>: tensor(7)
```
That is why I thought that it might be possible to make such function in C++. However, there's clearly a limitation on a traced function inputs. Hopefully @suo can help.

Reply:
Thanks for the code sample, I see what you mean now by "torch.Size.__getitem__ returns torch.Tensor with zero dimension IIRC". That is a dirty trick...
So I think that the regular limitations of tracing apply here: the if statements and for loop won't be properly considered :/

Reply:
It sounds to me like the easiest solution is to move this function into C++.

Reply:
On second thought, the most expedient solution is to settle on an interface with a workaround that creates a single-element temporary tensor. We can later move this to C++ if needed.